[RFC] - do not merge - edm::Wrapper<> de/serialisation

fwyzard opened the pull request:
Implement the ability to get, serialise, deserialise and put in the `Event` an arbitrary EDM product, handling it as an opaque type via its `edm::Wrapper<T>`.

This should work for products that do not make use of raw pointers or `edm::RefProd`/`edm::Ref`/`edm::Ptr`, etc.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-24389/6189

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-24389/6189/git-diff.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-24389/6189/git-diff.patch | patch -p1`

You can run `scram build code-checks` to apply code checks directly

fwyzard commented:
@cmsbuild, please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-24389/6190

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
A new Pull Request was created by  @fwyzard (Andrea Bocci) for master.

It involves the following packages:

HeterogeneousCore/MPICore


The following packages do not have a category, yet:

HeterogeneousCore/MPICore
Please create a PR for https://github.com/cms-sw/cms-bot/blob/master/categories_map.py to assign category

@cmsbuild can you please review it and eventually sign? Thanks.
@makortel this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
+1
Tested at: 0d2c02c0f7a39fb21006e9c3f0454961faae6a08
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24389/30084/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24389/30084/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3143859
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3143655
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

fwyzard commented:
@Dr15Jones does the `Provenance` information change during a job ?
If it doesn't, can I get it once (e.g. what `consumes` is declared) ?
Or do I have to get it at every event ?

Dr15Jones commented:
There are two parts to provenance. One is stable across the job once the work has been all setup (which would be around beginJob time). That includes all the information about configurations (it is only fully available at beginJob not construction time since 'fillDescriptions' can add additional parameters to the system). The second part can change each event as it records exactly what data products were actually requested by each module for that particular event.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-24389/6226

cmsbuild commented:
Pull request #24389 was updated. @cmsbuild can you please check and sign again.

Dr15Jones commented on HeterogeneousCore/MPICore/plugins/serialization.h:
```diff
@@ -0,0 +1,48 @@
+  std::unique_ptr<edm::WrapperBase> wrapper(reinterpret_cast<edm::WrapperBase *>(reinterpret_cast<char *>(buffer.ReadObjectAny(rootType)) + offset));
+  */
+  std::unique_ptr<edm::WrapperBase> wrapper(reinterpret_cast<edm::WrapperBase *>(buffer.ReadObjectAny(nullptr)));
+  return std::move(wrapper);
```
Best to not do a move since that stops an important compiler optimization. Instead you can do

return std::make_unique<edm::WrapperBase>(reinterpret_cast<edm::WrapperBase*>(buffer.ReadObjectAny(nullptr)));

fwyzard replied:
will do, thanks

davidlange6 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
There was an issue with git-cms-merge-topic you can see the log here: 
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24389/30150/git-merge-result

cmsbuild commented:
+1
Tested at: 7204e4400a3abc41a98cdd7192cb56ee57329b75
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24389/30163/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24389/30163/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3143879
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3143681
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-24389/6268

cmsbuild commented:
Pull request #24389 was updated. @cmsbuild can you please check and sign again.

davidlange6 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 6dd4201db2dcf5c94002ee9a3156a99fed78d3bd

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24389/30210/summary.html

I found follow errors while testing this PR

**Failed tests:  Build  ClangBuild**

* **Build**:

I found an error when building: 
 
 <pre>                 from /cvmfs/cms-ib.cern.ch/nweek-02539/slc6_amd64_gcc700/external/gcc/7.0.0-omkpbe2/include/c++/7.3.1/locale:43,
                 from /cvmfs/cms-ib.cern.ch/nweek-02539/slc6_amd64_gcc700/external/gcc/7.0.0-omkpbe2/include/c++/7.3.1/iomanip:43,
                 from /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_3_X_2018-09-01-1100/src/HeterogeneousCore/MPICore/plugins/GenericConsumer.cc:1:
/cvmfs/cms-ib.cern.ch/nweek-02539/slc6_amd64_gcc700/external/gcc/7.0.0-omkpbe2/include/c++/7.3.1/bits/unique_ptr.h: In instantiation of 'typename std::_MakeUniq<_Tp>::__single_object std::make_unique(_Args&& ...) [with _Tp = edm::WrapperBase; _Args = {edm::WrapperBase*}; typename std::_MakeUniq<_Tp>::__single_object = std::unique_ptr<edm::WrapperBase>]':
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_3_X_2018-09-01-1100/src/HeterogeneousCore/MPICore/plugins/serialization.h:44:112:   required from here
/cvmfs/cms-ib.cern.ch/nweek-02539/slc6_amd64_gcc700/external/gcc/7.0.0-omkpbe2/include/c++/7.3.1/bits/unique_ptr.h:825:30: error: invalid new-expression of abstract class type 'edm::WrapperBase'
     { return unique_ptr<_Tp>(new _Tp(std::forward<_Args>(__args)...)); }
                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /cvmfs/cms-ib.cern.ch/week1/slc6_amd64_gcc700/cms/cmssw-patch/CMSSW_10_3_X_2018-09-01-1100/src/DataFormats/Common/interface/Wrapper.h:10:0,
                 from /cvmfs/cms-ib.cern.ch/week1/slc6_amd64_gcc700/cms/cmssw-patch/CMSSW_10_3_X_2018-09-01-1100/src/DataFormats/Common/interface/ConvertHandle.h:6,
                 from /cvmfs/cms-ib.cern.ch/week1/slc6_amd64_gcc700/cms/cmssw-patch/CMSSW_10_3_X_2018-09-01-1100/src/FWCore/Framework/interface/Event.h:21,
</pre>

* **Clang**:

I found a compilation error while trying to compile with clang: 
I used this command: 
 scram b vclean && USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 16 COMPILER='llvm compile'


 <pre>>> Compiling edm plugin /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_3_X_2018-09-01-1100/src/HeterogeneousCore/MPICore/plugins/GenericConsumer.cc 
In file included from /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_3_X_2018-09-01-1100/src/HeterogeneousCore/MPICore/plugins/GenericConsumer.cc:1:
In file included from /cvmfs/cms-ib.cern.ch/nweek-02539/slc6_amd64_gcc700/external/gcc/7.0.0-omkpbe2/lib/gcc/x86_64-unknown-linux-gnu/7.3.1/../../../../include/c++/7.3.1/iomanip:43:
In file included from /cvmfs/cms-ib.cern.ch/nweek-02539/slc6_amd64_gcc700/external/gcc/7.0.0-omkpbe2/lib/gcc/x86_64-unknown-linux-gnu/7.3.1/../../../../include/c++/7.3.1/locale:43:
In file included from /cvmfs/cms-ib.cern.ch/nweek-02539/slc6_amd64_gcc700/external/gcc/7.0.0-omkpbe2/lib/gcc/x86_64-unknown-linux-gnu/7.3.1/../../../../include/c++/7.3.1/bits/locale_conv.h:41:
/cvmfs/cms-ib.cern.ch/nweek-02539/slc6_amd64_gcc700/external/gcc/7.0.0-omkpbe2/lib/gcc/x86_64-unknown-linux-gnu/7.3.1/../../../../include/c++/7.3.1/bits/unique_ptr.h:825:34: error: allocating an object of abstract class type 'edm::WrapperBase'
    { return unique_ptr<_Tp>(new _Tp(std::forward<_Args>(__args)...)); }
                                 ^
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_3_X_2018-09-01-1100/src/HeterogeneousCore/MPICore/plugins/serialization.h:44:15: note: in instantiation of function template specialization 'std::make_unique<edm::WrapperBase, edm::WrapperBase *>' requested here
  return std::make_unique<edm::WrapperBase>(reinterpret_cast<edm::WrapperBase *>(buffer.ReadObjectAny(nullptr)));
              ^
</pre>

cmsbuild commented:
Comparison not run due to Build errors (RelVals and Igprof tests were also skipped)

fwyzard committed:
edm::Wrapper<> de/serialisation

Implement the ability to get, serialise, deserialise and put in the Event an
arbitrary collection, handling it as an opaque type via its edm::Wrapper<T>.

fwyzard committed:
Add test producers and consumers

cmsbuild commented:
The code-checks are being triggered in jenkins.

fwyzard commented:
@cmsbuild, please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-24389/6269

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #24389 was updated. @cmsbuild can you please check and sign again.

cmsbuild commented:
+1
Tested at: 667f0a35c41a98105d6af028a17df3f10d2f097b
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24389/30211/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Pull request #24389 was updated. @makortel, @fwyzard can you please check and sign again.

silviodonato commented:
Hi @fwyzard , could you confirm that this PR is still needed/useful?

fwyzard commented:
Eventually, yes ...
